### PR TITLE
[Storage] KMS integration

### PR DESCRIPTION
**Pull request by frankyn**
Original author: @rossjudson
cc: @danoscarmike @rossjudson

Expected work items:
- [x] Define KMS keyring for integration tests
- [x] Allow configuration of default_kms_key_name in bucket
- [x] Allow specification of kms_key_name on object insert
- [x] Allow specification of kms_key_name on object rewrite
- [x] **Test**: Ensure kms_key_name is visible in object listings.
- [x] **Test**: Insert an object with a Cloud KMS key then fetch the object and verify contents and metadata (should include kms_key_name).
- [x] **Test**: Create a bucket with a default Cloud KMS key, insert an object, then fetch the object and verify its contents and metadata (should include kms_key_name).
- [x] **Test**: Update the default_kms_key_name of a bucket to a different key name.
- [x] **Test**: Update the bucket to remove default_kms_key_name encryption.
- [x] **Test**: Migrate CSEK to Cloud KMS key, download object and verify contents are decrypted.
- [x] **Test**: Test defaultKMSKeyName can be selected using `BucketField.ENCRYPTION`.
- [x] **Test**: Test kmsKeyName can be selected using `BlobField.KMS_KEY_NAME`.
- [x] **Test**: Write unit tests for BlobWriteOption and BlobTargetOption.
- [x] Nit: Remove Alpha and Beta annotations for GA features (Bucket level IAM and Requester Pays).
- [x] Add Beta annotations to KMS integration methods.
- [x] Code Review

### CODE REVIEW

File: google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java [hunk 1]
```diff
@@ -945,6 +970,10 @@ public DeleteRule apply(Rule rule) {
     if (billing != null) {
       builder.setRequesterPays(billing.getRequesterPays());
     }
+    Encryption encryption = bucketPb.getEncryption();
+    if (encryption != null && encryption.getDefaultKmsKeyName() != null && !encryption.getDefaultKmsKeyName().isEmpty()) {
+      builder.setDefaultKmsKeyName(encryption.getDefaultKmsKeyName());
```
Comment (pongad) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java]: Does Guava's `emptyToNull` work here?

File: google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java [hunk 2]
```diff
@@ -762,6 +778,13 @@ public StorageClass getStorageClass() {
     return labels;
   }
 
+  /**
+   * Returns the default Cloud KMS key to be applied to newly inserted objects in this bucket.
```
Comment (pongad) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java]: Is there an online doc the user can look up for details about what the KMS keys do?

Comment (frankyn) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java]: Not yet.

**Comment by pongad** [comment]
Also adding @garrettjonesgoogle

File: google-cloud-storage/pom.xml [hunk 3]
```diff
@@ -29,6 +29,7 @@
     <dependency>
       <groupId>com.google.apis</groupId>
       <artifactId>google-api-services-storage</artifactId>
+      <version>v1-rev125-1.23.0</version>
```
Comment (garrettjonesgoogle) [inline google-cloud-storage/pom.xml]: Don't put this version here, it should come from the bom.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Blob.java [hunk 4]
```diff
@@ -403,6 +402,12 @@ Builder setCustomerEncryption(CustomerEncryption customerEncryption) {
       return this;
     }
 
+    @Override
+    public Builder setKmsKeyName(String kmsKeyName) {
```
Comment (garrettjonesgoogle) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Blob.java]: Is KMS support GA?

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Bucket.java [hunk 5]
```diff
@@ -118,7 +118,6 @@ public static BucketSourceOption metagenerationNotMatch() {
      * Returns an option for blob's billing user project. This option is only used by the buckets with
      * 'requester_pays' flag.
      */
-    @GcpLaunchStage.Alpha
     public static BucketSourceOption userProject(String userProject) {
```
Comment (garrettjonesgoogle) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/Bucket.java]: Is the IAM stuff GA now?

File: google-cloud-storage/src/main/java/com/google/cloud/storage/spi/v1/HttpStorageRpc.java [hunk 6]
```diff
@@ -81,6 +81,8 @@
 import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
+
+import jdk.nashorn.internal.runtime.regexp.joni.exception.ValueException;
```
Comment (garrettjonesgoogle) [inline google-cloud-storage/src/main/java/com/google/cloud/storage/spi/v1/HttpStorageRpc.java]: Remove this import

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Blob.java (hunk 4 above)
Comment (frankyn) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/Blob.java]: Not yet, I'll expect merge at Beta. I'll annotate these methods with `@GcpLaunchStage.Beta`.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/Bucket.java (hunk 5 above)
Comment (frankyn) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/Bucket.java]: Yes. GA'd last year. Launch Id: 179764

Comment (frankyn) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/Bucket.java]: Requester pays as well. Launch Id: 191976

File: google-cloud-storage/src/test/java/com/google/cloud/storage/it/ITStorageTest.java [hunk 7]
```diff
@@ -204,6 +236,54 @@ public void testCreateBlobWithEncryptionKey() {
...
+    assertArrayEquals(BLOB_BYTE_CONTENT, readBytes);
+  }
+
+  @Test
+  public void testCreateBlobWithKmsKeyNameAndCustomerSuppliedKey() {
+    try {
+      String blobName = "test-create-with-kms-key-name-blob";
+      BlobInfo blob = BlobInfo.newBuilder(BUCKET, blobName).setKmsKeyName(KMS_KEY_NAME_1).build();
+      storage.create(blob, BLOB_BYTE_CONTENT, Storage.BlobTargetOption.encryptionKey(KEY));
+      fail("StorageException was expected"); // can't supply both.
```
[the remark this answers is not in the document]
Comment (frankyn) [reply google-cloud-storage/src/test/java/com/google/cloud/storage/it/ITStorageTest.java]: @garrettjonesgoogle and @pongad is this required? Other tests don't follow this pattern.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java (hunk 2 above)
Comment (frankyn) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java]: [Documentation](https://cloud.google.com/storage/docs/encryption/using-customer-managed-keys) is now available.

**Comment by frankyn** [comment]
@pongad and @garrettjonesgoogle, I realized yesterday afternoon that I confused myself on API surface for this feature in context of inserting an object with a KMS KeyName. On `objects.insert` kmsKeyName is set in request URL query parameters and only available in resource body post-operation. 

I updated the PR/design to use `BlobTargetOption` and `BlobWriteOption` and set `setKmsKeyName` in BlobInfo as a private method.

More context and background on the feature:
https://cloud.google.com/storage/docs/encryption/using-customer-managed-keys#add-object-key

**Comment by pongad** [comment]
Oh dear I'm sorry I dropped the ball on this. We moved a bunch of files. Could you merge from master?

There are some lint warnings that are worth fixing, but I'm fine fixing those in a separate PR.

File: google-cloud-clients/google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java [hunk 8]
```diff
@@ -584,6 +595,14 @@ public Builder setLabels(Map<String, String> labels) {
       return this;
     }
 
+    @GcpLaunchStage.Beta
+    @Override
+    public Builder setDefaultKmsKeyName(String defaultKmsKeyName) {
+      this.defaultKmsKeyName = defaultKmsKeyName != null
+              ? new String(defaultKmsKeyName) : Data.<String>nullOf(String.class);
```
[the remark this answers is not in the document]
Comment (pongad) [reply google-cloud-clients/google-cloud-storage/src/main/java/com/google/cloud/storage/BucketInfo.java]: I think this is worth fixing.

**Comment by pongad** [comment]
@garrettjonesgoogle Do you also want to take a look?

**Comment by frankyn** [comment]
All checks passed merging now.

**Pull request merged**